### PR TITLE
Add `Contract::finalize` entrypoint

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

An application's state was previously always loaded before and persisted after the execution of a cross-application call handler. This is inefficient in fuel consumption and also hurts execution performance in the validator.

## Proposal

Create a global variable with the application state inside the Wasm execution environment, and only load it once. This avoids reloading the state unnecessarily.

However, to persist the state only once, a new `finalize` entrypoint was added to the contract. It contains a default implementation that just persists the application state.

The new entrypoint also allows some new functionality, like having the option to discard some of the state changes at the end of the transaction, sending final messages, or canceling the execution, causing the block to be rejected. This last functionality also serves as a simpler alternative to sessions.

## Test Plan

Existing unit tests were updated to include testing the happy path of `finalize`, and some new unit tests were added to test if `finalize` can be used to reject blocks and send messages.

## Release Plan

Changes the WIT interface and the SDK APIs, so:
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.
- This PR is adding or removing Cargo features.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-execution/src/runtime.rs [hunk 1]
```diff
@@ -7,21 +7,24 @@ use crate::{
...
-    data_types::{Amount, ApplicationPermissions, ArithmeticError, BlockHeight, Timestamp},
+    data_types::{
+        Amount, ApplicationPermissions, ArithmeticError, BlockHeight, Resources, Timestamp,
+    },
     ensure,
     identifiers::{Account, ChainId, MessageId, Owner},
     ownership::ChainOwnership,
 };
 use linera_views::batch::Batch;
+use linked_hash_map::LinkedHashMap;
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:24]: I'm tempted to push back (again?) on adding this external library.
* External libraries add compile time and will need some security auditing down the road.
* This one says: "WARNING: THIS PROJECT IS IN MAINTENANCE MODE, DUE TO INSUFFICIENT MAINTAINER RESOURCES"
* Here, I believe we can just use `Vec` and `HashMap`.

File: linera-execution/src/runtime.rs [hunk 2]
```diff
@@ -953,64 +972,129 @@ impl ContractSyncRuntime {
...
-        }
-        runtime
-            .execution_outcomes
-            .push(ExecutionOutcome::User(application_id, outcome));
         Ok((runtime.execution_outcomes, runtime.resource_controller))
     }
+
+    /// Notifies all loaded applications that execution is finalizing.
+    fn finalize(&mut self, context: FinalizeContext) -> Result<(), ExecutionError> {
+        let applications = self
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:1012]: Are we supposed to set `self.is_finalizing` to true here?

Comment (jvff) [reply linera-execution/src/runtime.rs]: Oops! Yes, thanks! I fixed it and added some more unit tests for it.

File: linera-execution/src/lib.rs [hunk 3]
```diff
@@ -120,6 +120,14 @@ pub enum ExecutionError {
 
     #[error("Attempted to perform a reentrant call to application {0}")]
     ReentrantCall(UserApplicationId),
+    #[error(
+        "Application {callee_id} attempted to perform a cross-application to {callee_id} call \
```
Comment (afck) [inline linera-execution/src/lib.rs:124]:
```suggestion
        "Application {caller_id} attempted to perform a cross-application call to {callee_id} \
```

File: linera-sdk/src/lib.rs [hunk 4]
```diff
@@ -219,6 +219,19 @@ pub trait Contract: WithContractAbi + ContractAbi + Send + Sized {
         forwarded_sessions: Vec<SessionId>,
     ) -> Result<SessionCallOutcome<Self::Message, Self::Response, Self::SessionState>, Self::Error>;
 
+    /// Finishes the execution of the current transaction.
+    ///
+    /// This is called once before a transaction ends, to allow all applications that participated
+    /// in the transaction to perform any final operations, and optionally it may also cancel the
+    /// transaction if there are any pendencies.
```
Comment (afck) [inline linera-sdk/src/lib.rs:226]: This should probably explain that if you override it, you need to call `store` yourself, if you want to persist the contract state.

File: linera-execution/src/runtime.rs (hunk 1 above)
Comment (jvff) [reply linera-execution/src/runtime.rs]: Right! I had forgotten why we decided not to use it last time :sweat_smile:

File: linera-execution/src/lib.rs (hunk 3 above)
Comment (jvff) [reply linera-execution/src/lib.rs]: Oops, thanks for catching this!

File: linera-sdk/src/lib.rs (hunk 4 above)
Comment (jvff) [reply linera-sdk/src/lib.rs:226]: Good point. I tried to describe that, let me know what you think.

**Review by ma2bd — APPROVED** [review]

**Pull request merged**